tests/test_directive_code.py

```diff
@@ -11,7 +11,179 @@
+    content, lines = reader.read()
+    assert content == (u"# Literally included file using Python highlighting\n"
+                       u"# -*- coding: utf-8 -*-\n"
+                       u"\n"
+                       u"foo = \"Including Unicode characters: üöä\"\n")
+
+
+def test_LiteralIncludeReader_lines2():
+    options = {'lines': '1,4,6'}
+    reader = LiteralIncludeReader(LITERAL_INC_PATH, options, DUMMY_CONFIG)
+    content, lines = reader.read()
+    assert content == (u"# Literally included file using Python highlighting\n"
+                       u"foo = \"Including Unicode characters: üöä\"\n"
+                       u"class Foo:\n")
+
+
+@pytest.mark.sphinx()  # init locale for errors
+def test_LiteralIncludeReader_lines_and_lineno_match1(app, status, warning):
+    options = {'lines': '1,4,6', 'lineno-match': True}
+    reader = LiteralIncludeReader(LITERAL_INC_PATH, options, DUMMY_CONFIG)
+    with pytest.raises(ValueError):
+        content, lines = reader.read()
+
+
+@pytest.mark.sphinx()  # init locale for errors
+def test_LiteralIncludeReader_lines_and_lineno_match2(app, status, warning):
+    options = {'lines': '100-', 'lineno-match': True}
+    reader = LiteralIncludeReader(LITERAL_INC_PATH, options, DUMMY_CONFIG)
+    with pytest.raises(ValueError):
+        content, lines = reader.read()
+
+
+def test_LiteralIncludeReader_start_at():
+    options = {'start-at': 'Foo', 'end-at': 'Bar'}
+    reader = LiteralIncludeReader(LITERAL_INC_PATH, options, DUMMY_CONFIG)
+    content, lines = reader.read()
+    assert content == ("class Foo:\n"
+                       "    pass\n"
+                       "\n"
+                       "class Bar:\n")
+
+
+def test_LiteralIncludeReader_start_after():
+    options = {'start-after': 'Foo', 'end-before': 'Bar'}
+    reader = LiteralIncludeReader(LITERAL_INC_PATH, options, DUMMY_CONFIG)
+    content, lines = reader.read()
+    assert content == ("    pass\n"
+                       "\n")
+
+
+def test_LiteralIncludeReader_prepend():
+    options = {'lines': '1', 'prepend': 'Hello', 'append': 'Sphinx'}
+    reader = LiteralIncludeReader(LITERAL_INC_PATH, options, DUMMY_CONFIG)
+    content, lines = reader.read()
+    assert content == ("Hello\n"
+                       "# Literally included file using Python highlighting\n"
+                       "Sphinx\n")
+
+
+def test_LiteralIncludeReader_dedent():
+    # dedent: 2
+    options = {'lines': '10-12', 'dedent': 2}
+    reader = LiteralIncludeReader(LITERAL_INC_PATH, options, DUMMY_CONFIG)
+    content, lines = reader.read()
+    assert content == ("  def baz():\n"
+                       "      pass\n"
+                       "\n")
+
+    # dedent: 4
+    options = {'lines': '10-12', 'dedent': 4}
+    reader = LiteralIncludeReader(LITERAL_INC_PATH, options, DUMMY_CONFIG)
+    content, lines = reader.read()
+    assert content == ("def baz():\n"
+                       "    pass\n"
+                       "\n")
+
+    # dedent: 6
```

Comment:
I feel very strange that over dedent is allowed.
I think this should be warned.